File: app/src/main/java/com/example/merchio/RegisterActivity.java
package com.example.merchio;

import android.content.Intent;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Patterns;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.example.merchio.db.DbHelper;

public class RegisterActivity extends AppCompatActivity {

    EditText edtEmailRegister, edtPasswordRegister, edtUsername, edtConfirmPassword;
    Button btnRegister;
    TextView txtGoLogin;

    DbHelper dbHelper;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_register);

        dbHelper = new DbHelper(this);

        edtEmailRegister = findViewById(R.id.edtEmailRegister);
        edtPasswordRegister = findViewById(R.id.edtPasswordRegister);
        edtUsername = findViewById(R.id.edtUsername);
        edtConfirmPassword = findViewById(R.id.edtConfirmPassword);
        btnRegister = findViewById(R.id.btnRegister);
        txtGoLogin = findViewById(R.id.txtGoLogin);

        btnRegister.setOnClickListener(v -> registerUser());

        txtGoLogin.setOnClickListener(v -> {
            startActivity(new Intent(RegisterActivity.this, LoginActivity.class));
            finish();
        });
    }

    private void registerUser() {
        String email = edtEmailRegister.getText().toString().trim();
        String password = edtPasswordRegister.getText().toString().trim();
        String username = edtUsername.getText().toString().trim();
        String confirmPassword = edtConfirmPassword.getText().toString().trim();

        if (TextUtils.isEmpty(email)) {
            edtEmailRegister.setError("Email wajib diisi");
            edtEmailRegister.requestFocus();
            return;
        }

        if (!Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
            edtEmailRegister.setError("Format email tidak valid");
            edtEmailRegister.requestFocus();
            return;
        }

        if (TextUtils.isEmpty(password)) {
            edtPasswordRegister.setError("Password wajib diisi");
            edtPasswordRegister.requestFocus();
            return;
        }

        if (password.length() < 6) {
            edtPasswordRegister.setError("Password minimal 6 karakter");
            edtPasswordRegister.requestFocus();
            return;
        }

        if (TextUtils.isEmpty(username)) {
            edtUsername.setError("Username wajib diisi");
            edtUsername.requestFocus();
            return;
        }

        if (TextUtils.isEmpty(confirmPassword)) {
            edtConfirmPassword.setError("Konfirmasi password wajib diisi");
            edtConfirmPassword.requestFocus();
            return;
        }

        if (!password.equals(confirmPassword)) {
            edtConfirmPassword.setError("Password tidak sama");
            edtConfirmPassword.requestFocus();
            return;
        }

        if (dbHelper.isEmailExists(email)) {
            edtEmailRegister.setError("Email sudah terdaftar");
            edtEmailRegister.requestFocus();
            return;
        }

        // Karena form kamu belum punya field name, sementara name diisi sama dengan username.
        boolean success = dbHelper.insertUser(username, username, email, password);

        if (success) {
            Toast.makeText(this, "Register berhasil, silakan login", Toast.LENGTH_SHORT).show();

            Intent intent = new Intent(RegisterActivity.this, LoginActivity.class);
            startActivity(intent);
            finish();
        } else {
            Toast.makeText(this, "Register gagal", Toast.LENGTH_SHORT).show();
        }
    }
}

File: app/src/main/java/com/example/merchio/RegisterActivity.kt
package com.example.merchio

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import com.example.merchio.data.repository.AuthRepository
import com.example.merchio.utils.textString
import com.example.merchio.utils.toast
import android.widget.EditText
import kotlinx.coroutines.launch

class RegisterActivity : AppCompatActivity() {
    private val authRepository = AuthRepository()
    private lateinit var emailInput: EditText
    private lateinit var passwordInput: EditText
    private lateinit var usernameInput: EditText
    private lateinit var confirmInput: EditText
    private lateinit var registerButton: View
    private lateinit var loginText: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_register)
        emailInput = findViewById(R.id.edtEmailRegister)
        passwordInput = findViewById(R.id.edtPasswordRegister)
        usernameInput = findViewById(R.id.edtUsername)
        confirmInput = findViewById(R.id.edtConfirmPassword)
        registerButton = findViewById(R.id.btnRegister)
        loginText = findViewById(R.id.txtGoLogin)
        loginText.setOnClickListener { finish() }
        registerButton.setOnClickListener { register() }
    }

    private fun register() {
        val email = emailInput.textString()
        val username = usernameInput.textString()
        val password = passwordInput.textString()
        val confirm = confirmInput.textString()
        if (email.isBlank()) { emailInput.error = "Email wajib diisi"; return }
        if (username.isBlank()) { usernameInput.error = "Username wajib diisi"; return }
        if (password.length < 6) { passwordInput.error = "Minimal 6 karakter"; return }
        if (password != confirm) { confirmInput.error = "Password tidak sama"; return }

        registerButton.isEnabled = false
        lifecycleScope.launch {
            runCatching { authRepository.signUp(email, password, username) }
                .onSuccess {
                    toast("Register berhasil")
                    startActivity(Intent(this@RegisterActivity, MainActivity::class.java))
                    finishAffinity()
                }
                .onFailure { toast(it.message ?: "Register gagal") }
            registerButton.isEnabled = true
        }
    }
}
